8341696: C2: Non-fluid StringBuilder pattern bails out in OptoStringConcat

theoweidmannoracle opened the pull request:
Extends stringopts to also recognize non-fluid uses of StringBuilder and optimize them the same way.

For example, this basic case was not optimized before and is optimized with this PR:

```
StringBuilder sb = new StringBuilder();
sb.append("a");
sb.append(a);
return sb.toString();
```

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8341696](https://bugs.openjdk.org/browse/JDK-8341696): C2: Non-fluid StringBuilder pattern bails out in OptoStringConcat (**Enhancement** - P4)


### Reviewers
 * [Emanuel Peter](https://openjdk.org/census#epeter) (@eme64 - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/22537/head:pull/22537` \
`$ git checkout pull/22537`

Update a local copy of the PR: \
`$ git checkout pull/22537` \
`$ git pull https://git.openjdk.org/jdk.git pull/22537/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 22537`

View PR using the GUI difftool: \
`$ git pr show -t 22537`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/22537.diff">https://git.openjdk.org/jdk/pull/22537.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/22537#issuecomment-2516526634)
</details>

theoweidmannoracle commented on test/micro/org/openjdk/bench/vm/compiler/FluidSBBench.java:
```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
```
@shipilev I copied this benchmark from your RFE. Is this copyright notice correct?

shipilev replied:
Yes.

vnkozlov reviewed (approved):
Looks good.

TobiHartmann commented:
Please make sure to run testing with javac flag `-XDstringConcat` to have it use StringBuffer instead of invokedynamic based string concat (see [JEP 280](https://openjdk.org/jeps/280)).

eme64 commented on test/hotspot/jtreg/compiler/stringopts/TestFluidAndNonFluid.java:
```diff
@@ -0,0 +1,135 @@
+        Asserts.assertEQ("aZ", nonFluidOpaqueCall());
+    }
+
+    @Test
+    @IR(failOn = {IRNode.ALLOC_OF, "StringBuilder", IRNode.CALL_OF_METHOD, "toString", IRNode.INTRINSIC_TRAP})
+    public static String fluidNoParam() {
+        return new StringBuilder("0").append("a").append("c").toString();
+    }
```
Drive by comment: you seem to only have negative `failOn` tests here. You could consider adding a test where you have a positive rule, just to make sure you are matching the correct IR nodes.

eme64 replied:
Never mind, I did not see the positive cases below :see_no_evil:

theoweidmannoracle commented:
> Please make sure to run testing with javac flag `-XDstringConcat=inline` to have it use StringBuffer instead of invokedynamic based string concat (see [JEP 280](https://openjdk.org/jeps/280)).

As we discussed, this did not lead to any failures related to this change.

eme64 commented on test/hotspot/jtreg/compiler/stringopts/TestFluidAndNonFluid.java:
```diff
@@ -0,0 +1,135 @@
+ /*
+ * @test
+ * @bug 8341696
+ * @requires vm.compiler2.enabled
```
Not sure if I asked about this already: do we need this C2 restriction? The IR framework only checks IR  rules for C2, but the test could still do value verification for other settings where C2 is not available.

theoweidmannoracle replied:
Unintentional. Thanks for spotting it.

eme64 commented on src/hotspot/share/opto/stringopts.cpp:
```diff
@@ -408,27 +411,85 @@ Node_List PhaseStringOpts::collect_toString_calls() {
 }
 
-// Recognize a fluent-chain of StringBuilder/Buffer. They are either explicit usages
+PhaseStringOpts::CheckAppendResult PhaseStringOpts::check_append_candidate(CallStaticJavaNode* cnode,
```
Using two verbs in succession is a little confusing. At least write `check_and_append_candidate`. Or maybe `append_candidate_if_<something>`.

theoweidmannoracle replied:
It's not supposed to be a verb. This methods checks a potential `append` call, so it's an *append candidate*. Do you have any suggestions to clarify that?

eme64 replied:
Hmm... Ok. In my opinion a `check` method should be "pure", so it shoud not really have a side-effect... but you do some
`sc->add_control(cnode);` and `sc->push_string(arg);` etc.

I'm not familiar enough with the code yet, but those side-effects. Ah, the comment in the hpp seems to suggest you `add` it. So maybe:
`add_the_append_candidate_to_sc_if_<something>`
A bit long, but would be more helpful I think

theoweidmannoracle replied:
I think then I will just call it `process_append_candidate` then. There's no point in trying to explain the complex logic in the method name. This method together with its callee would better be moved into another class anyhow – but I gave up this idea because of all the codependencies.

eme64 replied:
Ok, so if I understand this right, this fully "processes" the `add`.
That is why you call it `CheckAppendResult`. Then maybe the enum tags could be a bit more descriptive...
`GoodAppend` -> `AddedAppendToStringConcat`
`GiveUp` implies that the algo outside is supposed to give up... but then it does continue to do something out there... so who is giving up? Maybe there could be a better name here.
`NotAppend` -> does this mean it is not an append, so you do nothing? -> `DidNothingBecauseNotAppend`...

You probably have even better ideas. Good names are hard :see_no_evil:

eme64 reviewed (changes requested):
Looks promising. Though I did not work with string-opts before, so a little hard for me to give a proper review. If you want/need me to review, a few more annotations with github comments would help. Otherwise I'll just leave it at the drive-by comments ;)

eme64 commented on src/hotspot/share/opto/stringopts.cpp:
```diff
@@ -543,7 +608,9 @@ StringConcat* PhaseStringOpts::build_candidate(CallStaticJavaNode* call) {
-          break;
+        } else if (use != nullptr &&
+                   check_append_candidate(use, sc, m, string_sig, int_sig, char_sig) == CheckAppendResult::GiveUp) {
+          return nullptr;
```
What happens here in the two other cases `GoodAppend` and `NotAppend`?

theoweidmannoracle replied:
We don't really care here if it was an append or not. The only important thing is to exit this algorithm if the processing of the append candidate detected a reason to definitely give up.

eme64 commented on src/hotspot/share/opto/stringopts.cpp:
```diff
@@ -569,62 +636,23 @@ StringConcat* PhaseStringOpts::build_candidate(CallStaticJavaNode* call) {
+      CheckAppendResult result = check_append_candidate(cnode, sc, m, string_sig, int_sig, char_sig);
+
+      if (result == CheckAppendResult::GiveUp) {
+        break;
```
Can you put a comment here where this is supposed to jump, and why?

theoweidmannoracle replied:
I was already considering to replace all the `break` with `return nullptr` here before, because all of them just exit from the entire algorithm because we detected something that completely prevents this optimization. But then I thought I will rather stick with the existing pattern and not change code that does not really need to be changed.

theoweidmannoracle replied:
Of course the way this entire loop works is a little confusing, but I didn't want to really refactor/rewrite all of this as it seems quite delicate and not trivial to test.

theoweidmannoracle commented:
@eme64 Thanks for your inputs. I added comments to document and explain the aspects you asked about, renamed symbols and replaced the breaks with more clear return statements. It should be clearer now hopefully.

theoweidmannoracle commented:
@eme64 @vnkozlov Could you take another look at the latest changes? They are mostly of cosmetic nature based on @eme64's suggestions.

eme64 commented on src/hotspot/share/opto/stringopts.cpp:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2009, 2023, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2009, 2024, Oracle and/or its affiliates. All rights reserved.
```
```suggestion
 * Copyright (c) 2009, 2025, Oracle and/or its affiliates. All rights reserved.
```

eme64 commented on src/hotspot/share/opto/stringopts.cpp:
```diff
@@ -408,27 +411,85 @@ Node_List PhaseStringOpts::collect_toString_calls() {
 //   .append("bar")
 //   .append(123)
-//   .toString(); // "foobar123"
+//   .toString(); // "foobar123"+
```
Is the "+" on purpose here?

theoweidmannoracle replied:
Thanks, for spotting this. Must be a typo

eme64 commented on src/hotspot/share/opto/stringopts.cpp:
```diff
@@ -489,7 +550,11 @@ StringConcat* PhaseStringOpts::build_candidate(CallStaticJavaNode* call) {
             use->method()->name() == ciSymbols::object_initializer_name() &&
             use->method()->holder() == m->holder()) {
-          // Matched the constructor.
+          //  Matched the constructor.
```
```suggestion
          // Matched the constructor.
```

eme64 commented on src/hotspot/share/opto/stringopts.cpp:
```diff
@@ -543,7 +608,15 @@ StringConcat* PhaseStringOpts::build_candidate(CallStaticJavaNode* call) {
+          // ...but we do not care if we really found an append or not:
+          // - If we found an append, that's perfect. Nothing further to do.
+          // - If this is a call to an unrelated method, validate_mem_flow() (and validate_control_flow())
+          //   will later check if this call prevents the optimization. So nothing to do here.
```
Ok, so why do we continue on and not return yet?

eme64 replied:
Maybe "We can keep looking for more appends"?

eme64 commented on src/hotspot/share/opto/stringopts.cpp:
```diff
@@ -408,27 +411,85 @@ Node_List PhaseStringOpts::collect_toString_calls() {
+        tty->cr();
+      }
+#endif
+      return ProcessAppendResult::AbortOptimization;
```
Question: is it a problem that we already did `sc->add_control(cnode);` in this case?

theoweidmannoracle replied:
No, since the entire StringConcat instance is thrown away, that is, never used. It is created in PhaseStringOpts::build_candidate which returns immediately if we return AbortOptimization.

eme64 commented on src/hotspot/share/opto/stringopts.hpp:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2009, 2023, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2009, 2024, Oracle and/or its affiliates. All rights reserved.
```
```suggestion
 * Copyright (c) 2009, 2025, Oracle and/or its affiliates. All rights reserved.
```

eme64 commented on test/hotspot/jtreg/compiler/stringopts/TestFluidAndNonFluid.java:
```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright (c) 2024 Oracle and/or its affiliates. All rights reserved.
```
```suggestion
 * Copyright (c) 2025 Oracle and/or its affiliates. All rights reserved.
```

eme64 commented on test/micro/org/openjdk/bench/vm/compiler/FluidSBBench.java:
```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+ * Copyright (c) 2024, Oracle and/or its affiliates. All rights reserved.
```
```suggestion
 * Copyright (c) 2025, Oracle and/or its affiliates. All rights reserved.
```

vnkozlov commented on src/hotspot/share/opto/stringopts.hpp:
```diff
@@ -53,6 +53,20 @@ class PhaseStringOpts : public Phase {
 
+  enum class ProcessAppendResult {
+    // Indicates that the candidate was indeed an append and process_append_candidate processed it
+    // accordingly (added it to the StringCocat etc.)
```
`StringCocat` -> `StringConcat`

vnkozlov commented on src/hotspot/share/opto/stringopts.hpp:
```diff
@@ -53,6 +53,20 @@ class PhaseStringOpts : public Phase {
+    AbortOptimization
+  };
+
+  // Called from build_candidate. Looks at an "append candidate", a call that might be a call to StringBuilder::append. If so, adds it to the StringConcat.
```
Please split line.

vnkozlov commented on src/hotspot/share/opto/stringopts.cpp:
```diff
@@ -569,62 +642,23 @@ StringConcat* PhaseStringOpts::build_candidate(CallStaticJavaNode* call) {
+        // some unhandled signature
 #ifndef PRODUCT
         if (PrintOptimizeStringConcat) {
-          tty->print("giving up because the call is effectively dead");
-          cnode->jvms()->dump_spec(tty); tty->cr();
+          tty->print("giving up because encountered unexpected signature ");
+          cnode->tf()->dump();
+          tty->cr();
+          cnode->in(TypeFunc::Parms + 1)->dump();
         }
 #endif
```
This print code could be put into `process_append_candidate()` too as you do for `AbortOptimization`.

theoweidmannoracle replied:
This here is printed if the return is `CandidateIsNotAppend` but we don't need to always print this, if the result is `CandidateIsNotAppend`. Further up, we don't care if the result is `CandidateIsNotAppend`, so I don't think it can be moved into `process_append_candidate`

vnkozlov replied:
ok

vnkozlov reviewed (approved):
Good

eme64 reviewed (approved):
Thanks for the updates @theoweidmannoracle ! Approved :)

theoweidmannoracle commented:
/integrate

vnkozlov reviewed (approved):
Re-approved after merge.

vnkozlov reviewed (changes requested):
@theoweidmannoracle, I see failures in your latest internal testing.

theoweidmannoracle commented:
> @theoweidmannoracle, I see failures in your latest internal testing.

Closed repo was out of sync. Latest testing has no new failures.

theoweidmannoracle commented:
/integrate

eme64 commented:
/sponsor
